include/micm/solver/backward_euler.inl

```diff
@@ -0,0 +1,195 @@
+        process_set.SubtractJacobianTerms(state.rate_constants_, Yn1, state.jacobian_);
+
+        // subtract the inverse of the time step from the diagonal
+        // TODO: handle vectorized jacobian matrix
+        for(auto& jac : state.jacobian_.AsVector()) {
+          jac *= -1;
+        }
+        for (std::size_t i_block = 0; i_block < state.jacobian_.Size(); ++i_block)
+        {
+          auto jacobian_vector = std::next(state.jacobian_.AsVector().begin(), i_block * state.jacobian_.FlatBlockSize());
+          for (const auto& i_elem : jacobian_diagonal_elements)
+            jacobian_vector[i_elem] -= 1 / H;
+        }
+
+        // We want to solve this equation for a zero
+        // (y_{n+1} - y_n) / H = f(t_{n+1}, y_{n+1})
+
+        // try to find the root by factoring and solving the linear system
+        linear_solver.Factor(state.jacobian_, state.lower_matrix_, state.upper_matrix_, singular);
+
+        auto yn1_iter = Yn1.begin();
+        auto yn_iter = Yn.begin();
+        auto forcing_iter = forcing.begin();
+        // forcing_blk in camchem
+        // residual = (Yn1 - Yn) / H - forcing;
+        // since forcing is only used once, we can reuse it to store the residual
+        for(; yn1_iter != Yn1.end(); ++yn1_iter, ++yn_iter, ++forcing_iter) {
+          *forcing_iter = (*yn1_iter - *yn_iter) / H - *forcing_iter;
+        }
+
+        // the result of the linear solver will be stored in forcing
+        // this represents the change in the solution
+        linear_solver.Solve(forcing, forcing, state.lower_matrix_, state.upper_matrix_);
+
+        // solution_blk in camchem
+        // Yn1 = Yn1 + residual;
+        // always make sure the solution is positive regardless of which iteration we are on
+        forcing_iter = forcing.begin();
+        yn1_iter = Yn1.begin();
+        for(; forcing_iter != forcing.end(); ++forcing_iter, ++yn1_iter) {
+          *yn1_iter = std::max(0.0, *yn1_iter + *forcing_iter);
+        }
+
+        // if this is the first iteration, we don't need to check for convergence
+        if (iterations++ == 0) continue;
+
+        // check for convergence
+        forcing_iter = forcing.begin();
+        yn1_iter = Yn1.begin();
+
+        // convergence happens when the absolute value of the change to the solution
+        // is less than a tolerance times the absolute value of the solution
+        do {
+          // changes that are much smaller than the tolerance are negligible and we assume can be accepted
+          converged = (std::abs(*forcing_iter) <= small) || (std::abs(*forcing_iter) <= tol * std::abs(*yn1_iter));
+          ++forcing_iter, ++yn1_iter;
+        }
+        while (converged && forcing_iter != forcing.end());
+
+        if (!converged) {
+          std::cout << "failed to converge within the newton iteration\n";
+        }
+      } while(!converged && iterations < max_iter);
+
+      if (!converged) {
+        std::cout << "failed to converge\n";
```

Comment:
If throwing an error, is this needed?

Reply:
Yes, we shouldn't throw an error; I placed that there to make sure we do whatever camchem does. A future PR will remove the error throwing and then at that time logging something would be appropriate, but maybe to `cerr`